### ISSUE TITLE
Diverging train and validation loss

### ISSUE THREAD

**Issue by wsnoble** [Noble-Lab/casanovo#226]
I thought I should see what happens on the dev branch using my M2 Mac. Rather than training on the full 9-species benchmark, I subsampled the first 120,693 spectra from the training set and the first 178,413 test set spectra.  It runs OK, but the training and validation loss immediately drop to really low values:

```
Step	Train loss	Valid loss	
3000	nan	0.008297
3772	0.816543	nan
6772	nan	0.001980
7544	0.002239	nan
10544	nan	0.000934
11316	0.001167	nan
14316	nan	0.001090
15088	0.000703	nan
18088	nan	0.000241
18860	0.000596	nan
21860	nan	0.000118
22632	0.000481	nan
25632	nan	0.000115
26404	0.000119	nan
29404	nan	0.000071
30176	0.000382	nan
33176	nan	0.000057
33948	0.000167	nan
```

Command line is
```
casanovo train target-head.train.mgf \
         --validation_peak_path target.test.20k.mgf \
         --config casanovo.yaml
```

I reproduced this on CentOS using a Quadro RTX4000, and now the validation losses go up

```
3772	0.816537	nan
6772	nan	6.458146
7544	0.002186	nan
 10544	nan	8.413699
 11316	0.001238	nan
 14316	nan	8.808524
 15088	0.000802	nan
 18088	nan	nan
 18860	0.000464	nan
 21860	nan	nan
```

Wout also reproduced the above behavior on his machine.

I am using  Casanovo version 3.3.1.dev13+gad48a09.

[casanovo.yaml.txt](https://github.com/Noble-Lab/casanovo/files/12232062/casanovo.yaml.txt)

Input files are too big to post but are available on [this thread](https://noblelab.slack.com/archives/C01MXN4NWMP/p1690821268922259?thread_ts=1690566461.416289&cid=C01MXN4NWMP),

**Comment by cfmelend** [comment]
During a training run with a multi-enzyme dataset using Casanovo version 3.3.1.dev13+gad48a09 (otherwise normal install according to the readthedocs instructions) I ran into the following similar very low train loss + nan's throughout the val loss:

```
 INFO [casanovo/MainProcess] model._log_history : Step	Train loss	Valid loss	
 INFO [casanovo/MainProcess] model._log_history : 10000	nan	7.928950
 INFO [casanovo/MainProcess] model._log_history : 20000	nan	nan
 INFO [casanovo/MainProcess] model._log_history : 28012	0.129454	nan
 INFO [casanovo/MainProcess] model._log_history : 38012	nan	nan
 INFO [casanovo/MainProcess] model._log_history : 48012	nan	nan
 INFO [casanovo/MainProcess] model._log_history : 56024	0.000228	nan
 INFO [casanovo/MainProcess] model._log_history : 66024	nan	nan
 INFO [casanovo/MainProcess] model._log_history : 76024	nan	nan
 INFO [casanovo/MainProcess] model._log_history : 84036	0.000638	nan
 INFO [casanovo/MainProcess] model._log_history : 94036	nan	nan
 INFO [casanovo/MainProcess] model._log_history : 104036	nan	nan
 INFO [casanovo/MainProcess] model._log_history : 112048	0.000946	nan
 INFO [casanovo/MainProcess] model._log_history : 122048	nan	nan
 INFO [casanovo/MainProcess] model._log_history : 132048	nan	nan
 INFO [casanovo/MainProcess] model._log_history : 140060	0.001395	nan
 INFO [casanovo/MainProcess] model._log_history : 150060	nan	nan
 INFO [casanovo/MainProcess] model._log_history : 160060	nan	nan
 INFO [casanovo/MainProcess] model._log_history : 168072	0.001183	nan
 INFO [casanovo/MainProcess] model._log_history : 178072	nan	nan
 INFO [casanovo/MainProcess] model._log_history : 188072	nan	nan
 INFO [casanovo/MainProcess] model._log_history : 196084	0.000320	nan
 INFO [casanovo/MainProcess] model._log_history : 206084	nan	nan
 INFO [casanovo/MainProcess] model._log_history : 216084	nan	nan
 INFO [casanovo/MainProcess] model._log_history : 224096	0.000349	nan
 INFO [casanovo/MainProcess] model._log_history : 234096	nan	nan
 INFO [casanovo/MainProcess] model._log_history : 244096	nan	nan
 INFO [casanovo/MainProcess] model._log_history : 252108	0.000306	nan
 INFO [casanovo/MainProcess] model._log_history : 262108	nan	nan
 INFO [casanovo/MainProcess] model._log_history : 272108	nan	nan
 INFO [casanovo/MainProcess] model._log_history : 280120	0.000272	nan
 INFO [casanovo/MainProcess] model._log_history : 290120	nan	nan

```

**Comment by wsnoble** [comment]
Carlo, have you tried this using a previous version on the dev branch?

**Comment by cfmelend** [comment]
Not with this dataset. I was able to train on a subset of the MassIVE-KB benchmark dataset (~50,000 spectra) on a slightly older version of the dev branch (Casanovo version 3.3.1.dev8+g6299bd2) without seeing this pattern of very low train/val loss or any kind of nan's in the outputs.

**Comment by cfmelend** [comment]
Following up on suggestions that this might have been caused by the dataloaders failing to properly shuffle training and/or validation batches I added a print statement to the end of [this function](https://github.com/Noble-Lab/casanovo/blob/main/casanovo/denovo/dataloaders.py#L162). The printouts of the batch items for both train and valsets don't appear to be identical or repeating (I attached these here in case someone spots something I don't). This test was performed with depthcharge-ms==0.2.2 and Casanovo version 3.3.1.dev13+gad48a09.d20230808 on an A100 node.
[dataloader_shuffle_test.txt](https://github.com/Noble-Lab/casanovo/files/12296393/dataloader_shuffle_test.txt)
